sources/editor/Stride.Assets.Presentation/ViewModel/CodeViewModel.cs

```diff
@@ -99,36 +100,9 @@ void TrackedAssemblies_CollectionChanged(object sender, Core.Collections.Trackin
                 // TODO: Right now, we simply replace the solution with newly loaded one
                 // Ideally, we should keep our existing solution and update it to follow external changes after initial loading (similar to VisualStudioWorkspace)
                 // This should provide better integration with background changes and local changes
-                projectWatcher.AssembliesChangedBroadcast.LinkTo(new ActionBlock<List<AssemblyChangedEvent>>(events =>
-                {
-                    if (events.Count == 0)
-                        return;
-
-                    Dispatcher.InvokeAsync(async () =>
-                    {
-                        // Update projects
-                        foreach (var e in events.Where(x => x.ChangeType == AssemblyChangeType.Project))
-                        {
-                            var project = e.Project;
-                            if (project != null)
-                            {
-                                await ReloadProject(strideAssetsViewModel.Session, project);
-                            }
-                        }
-
-                        // Update files
-                        foreach (var e in events.Where(x => x.ChangeType == AssemblyChangeType.Source))
-                        {
-                            var documentId = workspace.CurrentSolution.GetDocumentIdsWithFilePath(e.ChangedFile).FirstOrDefault();
-                            if (documentId != null)
-                                workspace.HostDocumentTextLoaderChanged(documentId, new FileTextLoader(e.ChangedFile, null));
-                        }
-                    }).Wait();
-                }), new DataflowLinkOptions());
-
+                ProjectUpdate = UpdateProjects(projectWatcher, workspace, strideAssetsViewModel);
                 return workspace;
```

Comment:
At this point, we are returned a workspace that might not be up to date yet (the async loop in UpdateProjects might not have completed yet). It feels very unstable (the previous code was like that as well).

What if instead, we make the whole continuation (from line 64) asynchronous as well so that we ensure that `workspaceTask` (and thus the `Workspace` property from line 152) only completes when everything is ready?

The code would look like this:

```csharp
// line 64
workspaceTask = workspaceTask.ContinueWith(async previousTask  =>
{
    var projectWatcher = await projectWatcherTask;
    var workspace = await previousTask;

    // code unchanged

    // line 103
    await UpdateProjects(projectWatcher, workspace, strideAssetsViewModel);
    return workspace;
}
```